src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/PENamedTypeSymbol.vb

```diff
@@ -1556,6 +1557,94 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
         Friend NotOverridable Overrides Function GetSynthesizedWithEventsOverrides() As IEnumerable(Of PropertySymbol)
             Return SpecializedCollections.EmptyEnumerable(Of PropertySymbol)()
         End Function
+
+        Private ReadOnly Property HasAnyDeclaredRequiredMembers As Boolean
+            Get
+                Return ContainingPEModule.Module.HasAttribute(Handle, AttributeDescription.RequiredMemberAttribute)
+            End Get
+        End Property
+
+        Private Shared ReadOnly s_requiredMembersErrorSentinel As ImmutableSegmentedDictionary(Of String, Symbol) = ImmutableSegmentedDictionary(Of String, Symbol).Empty.Add("<error sentinel>", Nothing)
+        ''' <summary>
+        ''' <see langword="Nothing" /> if uninitialized. <see cref="s_requiredMembersErrorSentinel"/> if there are errors.
+        ''' <see cref="ImmutableSegmentedDictionary(Of String, Symbol).Empty"/> if there are no required members. Otherwise,
+        ''' the required members.
+        ''' </summary>
+        Private _lazyRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol) = Nothing
+
+        Friend Overrides ReadOnly Property AllRequiredMembers As ImmutableSegmentedDictionary(Of String, Symbol)
+            Get
+                CalculateRequiredMembersIfRequired()
+                Debug.Assert(Not _lazyRequiredMembers.IsDefault)
+                Return _lazyRequiredMembers
+            End Get
+        End Property
+
+        Friend Overrides ReadOnly Property HasRequiredMembersError As Boolean
+            Get
+                CalculateRequiredMembersIfRequired()
+                Debug.Assert(Not _lazyRequiredMembers.IsDefault)
+                Return _lazyRequiredMembers = s_requiredMembersErrorSentinel
+            End Get
+        End Property
+
+        Private Sub CalculateRequiredMembersIfRequired()
+            If Not _lazyRequiredMembers.IsDefault Then
+                Return
+            End If
+
+            Dim builder As ImmutableSegmentedDictionary(Of String, Symbol).Builder = Nothing
+            Dim success = TryCalculateRequiredMembers(builder)
+
+            Dim requiredMembers = If(success,
+                If(builder?.ToImmutable(), ImmutableSegmentedDictionary(Of String, Symbol).Empty),
+                s_requiredMembersErrorSentinel)
+
+            RoslynImmutableInterlocked.InterlockedInitialize(_lazyRequiredMembers, requiredMembers)
+        End Sub
+
+        Private Function TryCalculateRequiredMembers(ByRef requiredMembersBuilder As ImmutableSegmentedDictionary(Of String, Symbol).Builder) As Boolean
+            Dim lazyRequiredMembers = _lazyRequiredMembers
+            If lazyRequiredMembers = s_requiredMembersErrorSentinel Then
```

Comment:
>If lazyRequiredMembers = s_requiredMembersErrorSentinel Then [](http://example.com/codeflow?start=12&length=60)

We checked ```_lazyRequiredMembers``` before calling this method. Why doing this again?

Reply:
This is a recursive function, base types should check this.

Reply:
> This is a recursive function, base types should check this.

I do not think we can rely on this recursion for bases, because base is not always a PENAmedTypeSymbol. We should request the information from AllRequiredMembers property instead.

Reply:
I was trying to avoid putting `_lazyRequiredMembers` on all `NamedTypeSymbol`s in VB, but I think you've convinced me I can't avoid it. I can certainly make it a bit smarter about caching the dictionary as it goes along.